integrationtests/self/http_test.go

```diff
@@ -1068,4 +1068,105 @@ var _ = Describe("HTTP tests", func() {
 			"Unannounced": {"Surprise!"},
 		})))
 	})
+
+	It("graceful shutdown successfully with no timeout", func() {
+		var (
+			startShutdownChan   = make(chan struct{})
+			shutdownStartedChan = make(chan struct{})
+			clientDoneChan      = make(chan struct{})
+			serverDoneChan      = make(chan struct{})
+		)
+
+		mux.HandleFunc("/slow", func(w http.ResponseWriter, r *http.Request) {
+			close(startShutdownChan)
+			<-shutdownStartedChan
+			chunkSize := len(PRDataLong) / 10
+			for i := range 10 {
+				w.Write(PRDataLong[i*chunkSize : (i+1)*chunkSize])
+			}
+		})
+		go func() {
+			resp, err := client.Get(fmt.Sprintf("https://localhost:%d/slow", port))
+			Expect(err).ToNot(HaveOccurred())
+			Expect(resp.StatusCode).To(Equal(200))
+			body, err := io.ReadAll(resp.Body)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(body).To(Equal(PRDataLong))
+			close(clientDoneChan)
+		}()
+
+		go func() {
+			<-startShutdownChan
+			close(shutdownStartedChan)
+			err := server.CloseGracefully(context.Background())
+			Expect(err).ToNot(HaveOccurred())
+			close(serverDoneChan)
+		}()
+		Eventually(clientDoneChan, "5s").Should(BeClosed())
+		Eventually(serverDoneChan, "5s").Should(BeClosed())
+	})
+
+	It("graceful shutdown successfully with timeout", func() {
+		// server will begin shutdown after receiving two requests
+		var (
+			fastChan       = make(chan struct{})
+			slowChan       = make(chan struct{})
+			fastDoneChan   = make(chan struct{})
+			slowDoneChan   = make(chan struct{})
+			serverDoneChan = make(chan struct{})
+		)
+
+		mux.HandleFunc("/fast", func(w http.ResponseWriter, r *http.Request) {
+			close(fastChan)
+			w.Write(PRDataLong)
```

Comment:
Write less data.

Reply:
By reducing `MaxIdleTime`, the time to finish a test can be reduced to a degree, but not that small. It can't be reduced to less than half a second.

Reply:
Of course it can, and I already suggested you how to do it:
1. Call `CloseGracefully`.
2. After 150ms, cancel the context.
3. After 300ms, assert that the request has returned.

Reply:
What do you think that I wrap the `http.ResponseWriter` that will extend the deadline everytime data is written. It will also embed a client context that will be cancelled in case the deadline exceeded.

I feel like the with timeout test should have some tolerance for a longer test. After all, it's simulating a long request interrupted. I'll keep it under 10 seconds.

Reply:
> What do you think that I wrap the http.ResponseWriter that will extend the deadline everytime data is written. It will also embed a client context that will be cancelled in case the deadline exceeded.

I don't understand what this would test.

> I feel like the with timeout test should have some tolerance for a longer test. After all, it's simulating a long request interrupted. I'll keep it under 10 seconds.

The entire HTTP test suite, all 35 integration tests, runs in less than 2s. It is not reasonable to add a single test that runs 5x as long.

Reply:
> > What do you think that I wrap the http.ResponseWriter that will extend the deadline everytime data is written. It will also embed a client context that will be cancelled in case the deadline exceeded.
> 
> I don't understand what this would test.
> 
> >

Not a test, but the implementation detail that's used to test. Because when the server is closed and ongoing requests are supposedly interrupted, quic needs some time to react to the loss of connectivity. Using context and deadline we can stop when no progress is made.



> > I feel like the with timeout test should have some tolerance for a longer test. After all, it's simulating a long request interrupted. I'll keep it under 10 seconds.
> 
> The entire HTTP test suite, all 35 integration tests, runs in less than 2s. It is not reasonable to add a single test that runs 5x as long.

How about 2.5x as long as a compromise. I could try for lower but that just magic numbers from then on.

Reply:
> Because when the server is closed and ongoing requests are supposedly interrupted, quic needs some time to react to the loss of connectivity. Using context and deadline we can stop when no progress is made.

It doesn't. We're not testing a connection timeout. When the context of `CloseGracefully` is canceled, the server is supposed to send a CONNECTION_CLOSE, which immediately closes the connection. 300ms, as I suggested earlier, should be **plenty** of time to test this.

Reply:
The current implementation doesn't close those quic connections. The only thing it closes is the underlying datagram socket. Hence the use of context.

I suppose it can be added here? The signature needs to be changed.

![image](https://github.com/user-attachments/assets/260dfa32-46ff-4939-a912-230e3e3aaa30)

Reply:
But then it only sends a CONNECTION_CLOSE, when all serve routine is done. It should send CONNECTION_CLOSE when the context passed to `CloseGracefully` is done.

Reply:
Perhaps we need to store quic connections in a map?

Reply:
> Perhaps we need to store quic connections in a map?

There's probably a better solution by canceling a context, closing a channel or something along those lines.

Reply:
By using channel, a new goroutine is required. Also the main closing call needs to know if these `CONNECTION_CLOSE`s finish sending,

Reply:
In the first draft I'll use map to store these connections. Will come back later to see the alternative implementations.

Reply:
Perhaps, I can use another context [here](https://github.com/quic-go/quic-go/blob/master/http3/server.go#L490) that will be responsible for closing the connection. This context is tied with another context associated with server that will be canceled when `CONNECTION_CLOSE` should be sent. Yet another counter and channel. Any idea why the context passed isn't used though?

Reply:
@marten-seemann Is it safe to use another context [here](https://github.com/quic-go/quic-go/blob/master/http3/conn.go#L181)? It's shared by both server and client. If no map is used, this is the best place I can think of.

Reply:
You could pass a context to `HandleUnidirectionalStreams`, right?

Reply:
No, you can't

Reply:
What's with the snarky answers?

Do you want this PR to get merged or not? If you're not interested in resolving the remaining issues, shall we just close this PR and not waste any more of both of our time?

Reply:
I'm not snarky. The [function](https://github.com/quic-go/quic-go/blob/master/http3/conn.go#L173) signature is `HandleUnidirectionalStreams(hijack func(StreamType, quic.ConnectionTracingID, quic.ReceiveStream, error) (hijacked bool))`. There is no way to pass a context.

Reply:
Can I use some of the codes [here](https://github.com/quic-go/quic-go/pull/4689) in this branch instead? Some of the changes conflict.

Reply:
I suggest we merge #4689 first. This is a bug.